src/hardware/affinity.cpp

```diff
@@ -51,10 +52,14 @@ std::vector<int> get_affinity() {
 }
 #endif
 
-unsigned count_available_cores() {
-    return get_affinity().size();
+util::optional<std::size_t> num_available_cores() {
```

Comment:
Just on naming: namespace `hw` is great, but why not also `src/hw/...` for the code? It's still clear and unambiguous IMO, and has the namespace and directory name line up, which is convenient.

`num_available_cores()`: why not just `num_cores()` or `count_cores()`? The latter might be clearer in that it indicates that it is an active process rather than simply looking something up.